src/map/unit.cpp

```diff
@@ -1840,30 +1847,109 @@ void unit_set_attackdelay(block_list& bl, t_tick tick, e_delay_event event)
 	if (ud == nullptr)
 		return;
 
+	uint16 attack_delay = 0;
+	uint16 act_delay = 0;
+
 	switch (bl.type) {
 		case BL_PC:
 			switch (event) {
 				case DELAY_EVENT_ATTACK:
+				case DELAY_EVENT_CASTBEGIN_ID:
+				case DELAY_EVENT_CASTBEGIN_POS:
 				case DELAY_EVENT_PARRY:
-					// TODO: This should also happen on cast begin
 					// Officially for players it just remembers the last attack time here and applies the delays during the comparison
 					// But we pre-calculate the delays instead and store them in attackabletime and canact_tick
-					ud->attackabletime = tick + status_get_adelay(&bl);
+					attack_delay = status_get_adelay(&bl);
 					// A fixed delay is added here which is equal to the minimum attack motion you can get
 					// This ensures that at max ASPD attackabletime and canact_tick are equal
-					ud->canact_tick = tick + status_get_amotion(&bl) + (pc_maxaspd(reinterpret_cast<map_session_data*>(&bl)) / AMOTION_DIVIDER_PC);
+					act_delay = status_get_amotion(&bl) + (pc_maxaspd(reinterpret_cast<map_session_data*>(&bl)) / AMOTION_DIVIDER_PC);
 					break;
 			}
 			break;
-		default:
+		case BL_MOB:
+			switch (event) {
+				case DELAY_EVENT_ATTACK:
+					// This represents setting of attack delay (recharge time) that happens for non-PCs
+					attack_delay = status_get_adelay(&bl);
+					break;
+			}
+			break;
+		case BL_HOM:
+			switch (event) {
+				case DELAY_EVENT_ATTACK:
+					// This represents setting of attack delay (recharge time) that happens for non-PCs
+					attack_delay = status_get_adelay(&bl);
+					break;
+				case DELAY_EVENT_CASTBEGIN_ID:
+				case DELAY_EVENT_CASTBEGIN_POS:
+					// For non-PCs that can be controlled from the client, there is a security delay of 200ms
+					// However to prevent tricks to use skills faster, we have a config to use amotion instead
+					if (battle_config.amotion_min_skill_delay == 1 && status_get_amotion(&bl) > MIN_DELAY_SLAVE)
+						act_delay = status_get_amotion(&bl);
```

Comment:
```suggestion
					if (battle_config.amotion_min_skill_delay == 1)
						act_delay = status_get_amotion(&bl) + MAX_ASPD_NOPC;
```

Reply:
Do we really want to increase the delay by MAX_ASPD_NOPC? Also here you are making the assumption that MIN_DELAY_SLAVE is always MAX_ASPD_NOPC*2, which only works until someone customizes that value.

Reply:
When that setting is set, I would harmonize the behavior.
Like you did it for BL_MER in DELAY_EVENT_CASTBEGIN_POS